File: packages/folio-core/domain/knowledge.go
package domain

import "time"

// Knowledge is a durable note that outlives the project it was learned on: a
// tip, a snippet, a fix worth keeping, a page worth not losing.
//
// It sits outside the tenancy every other record obeys. Any authenticated user
// reads and writes all of it, because the things worth keeping here are things
// like how to get realtime working behind a proxy, and copying that into each
// project would be the only alternative.
type Knowledge struct {
	ID KnowledgeID
	// ProjectID is optional and is a note of where this was learned, not a
	// fence around who may read it. A project's deletion detaches its
	// knowledge rather than destroying it.
	ProjectID ProjectID
	Slug      string
	Title     string
	Body      string
	Tags      []string
	// CreatedBy is always set, though nothing filters on it: attribution here
	// tells a reader whom to ask, it does not decide who may look.
	CreatedBy UserID
	CreatedAt time.Time
	UpdatedAt time.Time
}

// KnowledgeFilter narrows a listing. Zero values mean "no restriction", and an
// unset ProjectID therefore lists everything rather than only the unattached
// notes; Unattached asks for those.
type KnowledgeFilter struct {
	ProjectID  ProjectID
	Unattached bool
	Tags       []string
	Search     string
	Limit      int
	Offset     int
}

File: apps/web/src/core/ports/knowledge.ts
import type { Result } from '_/lib/result'

// Knowledge is durable, reusable notes: a tip, a snippet, a fix worth keeping.
// It is the one resource with no project, so unlike every other port here
// nothing it takes is scoped to one.
export type Knowledge = {
	readonly id: string
	readonly slug: string
	readonly title: string
	readonly body: string
	// Empty when the note belongs to no project, which is the common case. It
	// records where the note was learned and does not restrict who may read it.
	readonly projectId: string
	readonly tags: readonly string[]
	readonly createdAt: Date
	readonly updatedAt: Date
}

export type KnowledgeFilter = {
	readonly text?: string
	readonly tags?: readonly string[]
	readonly limit?: number
}

export type KnowledgePort = {
	readonly list: (filter?: KnowledgeFilter) => Promise<Result<readonly Knowledge[]>>
	// Takes an id or a slug: the slug namespace is global, so no project is
	// needed to resolve one.
	readonly get: (ref: string) => Promise<Result<Knowledge>>
}
